runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/PipelineTranslation.java

```diff
@@ -102,11 +121,370 @@ public void visitPrimitiveTransform(Node node) {
+        inputReplacements.put(
+            originalOutputs.get(originalOutputs.keySet().iterator().next()).getName(),
+            ((PCollection) output).getName());
+      } else {
+        throw new RuntimeException("Unexpected output type");
+      }
+
+      // We create a new AppliedPTransform to represent the upgraded transform and register it in an
+      // SdkComponents object.
+      AppliedPTransform<?, ?, ?> updatedAppliedPTransform =
+          AppliedPTransform.of(
+              appliedPTransform.getFullName() + "_external",
+              appliedPTransform.getInputs(),
+              newOutputs,
+              externalTransform,
+              externalTransform.getResourceHints(),
+              appliedPTransform.getPipeline());
+      SdkComponents updatedComponents =
+          SdkComponents.create(
+              runnerAPIpipeline.getComponents(), runnerAPIpipeline.getRequirementsList());
+      String updatedTransformId =
+          updatedComponents.registerPTransform(updatedAppliedPTransform, Collections.emptyList());
+      RunnerApi.Components updatedRunnerApiComponents = updatedComponents.toComponents();
+
+      // Recording input updates to the transforms to refer to the upgraded transform instead of the
+      // old one.
+      // Also recording the newly generated id of the old (overridden) transform in the
+      // updatedRunnerApiComponents.
+      Map<String, Map<String, String>> transformInputUpdates = new HashMap<>();
+      List<String> oldTransformIds = new ArrayList<>();
+      updatedRunnerApiComponents
+          .getTransformsMap()
+          .forEach(
+              (transformId, transform) -> {
+                // Mapping from existing key to new value.
+                Map<String, String> updatedInputMap = new HashMap<>();
+                for (Map.Entry<String, String> entry : transform.getInputsMap().entrySet()) {
+                  if (inputReplacements.containsKey(entry.getValue())) {
+                    updatedInputMap.put(entry.getKey(), inputReplacements.get(entry.getValue()));
+                  }
+                }
+                for (Map.Entry<String, String> entry : transform.getOutputsMap().entrySet()) {
+                  if (inputReplacements.containsKey(entry.getValue())
+                      && urn.equals(transform.getSpec().getUrn())) {
+                    oldTransformIds.add(transformId);
+                  }
+                }
+                if (updatedInputMap.size() > 0) {
+                  transformInputUpdates.put(transformId, updatedInputMap);
+                }
+              });
+      // There should be only one recorded old (upgraded) transform.
+      if (oldTransformIds.size() != 1) {
+        throw new IOException(
+            "Expected exactly one transform to be updated by "
+                + oldTransformIds.size()
+                + " were updated.");
+      }
+      String oldTransformId = oldTransformIds.get(0);
```

Comment:
Shouldn't we already know exactly what transform we're updating? This seems a round-about way of getting it, and will break if a transform doesn't have outputs.